plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,611 @@
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_reset (FuPxiDevice *self, GError **error)
+{
+	guint8 req[FU_PXI_DEVICE_OTA_BUF_SZ] = {
+		PXI_HID_DEV_OTA_OUTPUT_REPORT_ID,
+		FU_PXI_DEVICE_CMD_FW_MCU_RESET,
+		OTA_RESET,
+	};
+	fu_device_set_status (FU_DEVICE (self), FWUPD_STATUS_DEVICE_RESTART);
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0,
+					 req, 0x3,
+					 error)) {
+		g_prefix_error (error, "failed to reset: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_fw_ota_init (FuPxiDevice *self, GError **error)
+{
+	const guint8 req[] = {
+		PXI_HID_DEV_OTA_OUTPUT_REPORT_ID,
+		FU_PXI_DEVICE_CMD_FW_OTA_INIT,
+	};
+	return fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0,
+					   req, sizeof(req), error);
+}
+
+static gboolean
+fu_pxi_device_fw_ota_init_new (FuPxiDevice *self, gsize bufsz, GError **error)
+{
+	guint8 res[FU_PXI_DEVICE_OTA_BUF_SZ] = { 0x0 };
+	guint8 fw_version[10] = { 0x0 };
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+
+	/* write fw ota init new command */
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_FEATURE_REPORT_ID);
+	fu_byte_array_append_uint8 (req, FU_PXI_DEVICE_CMD_FW_OTA_INIT_NEW);
+	fu_byte_array_append_uint32 (req, bufsz, G_LITTLE_ENDIAN);
+	fu_byte_array_append_uint8 (req, 0x0);	/* OTA setting */
+	g_byte_array_append (req, fw_version, sizeof(fw_version));
+	if (!fu_pxi_device_set_feature (self, req->data, req->len, error))
+		return FALSE;
+
+	/* delay for BLE device read command */
+	g_usleep (10 * 1000);
+
+	/* read fw ota init new command */
+	res[0] = PXI_HID_DEV_OTA_FEATURE_REPORT_ID;
+	res[1] = FU_PXI_DEVICE_CMD_FW_OTA_INIT_NEW;
+	if (!fu_pxi_device_get_feature (self, res, sizeof(res), error))
+		return FALSE;
+
+	/* shared state */
+	if (!fu_common_read_uint8_safe (res, sizeof(res), 0x3,
+					&self->status, error))
+		return FALSE;
+	if (!fu_common_read_uint8_safe (res, sizeof(res), 0x4,
+					&self->new_flow, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0x5,
+					 &self->offset, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0x7,
+					 &self->checksum, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint32_safe (res, sizeof(res), 0x9,
+					 &self->max_object_size, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0xd,
+					 &self->mtu_size, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0xf,
+					 &self->prn_threshold, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint8_safe (res, sizeof(res), 0x11,
+					&self->spec_check_result, error))
+		return FALSE;
+
+	/* sanity check */
+	if (self->spec_check_result != OTA_SPEC_CHECK_OK) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "FwInitNew Spec Check Fail 0x%02x",
+			     self->spec_check_result);
```

Comment:
assuming the to string helper I mentioned gets made, it should be used here to.

Reply:
Fixed.